### PR TITLE
HIVE-27760: Filter on date type partitioning column producing 0 results

### PR DESCRIPTION
**Pull request by mdayakar**
### What changes were proposed in this pull request?
**RootCause:** As a part of [HIVE-27373](http://example.com/) issue fix to parse the string to java.sql.Date object, java.text.SimpleDateFormat is replaced with java.time.format.DateTimeFormatter using java.time.LocalDate which represents a Date without TimeZone.  Here this input is passed [here ](https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java#L1370) which uses SimpleDateFormat(parsing dates in a locale-sensitive manner) and java.sql.Date. Here user timezone is passed different so actual value is getting changed to a different value (for example 2023-01-01 is changed to 2022-12-31) which is not matching with any partition so nothing gets returned.

**Solution:** In MetaStoreDirectSql.java, we should use java.time.format.DateTimeFormatter with java.time.LocalDate so that it will return proper date string.

### Why are the changes needed?
To fix HIVE-27760 otherwise it will results-in wrong results when Date/Timestamp partition column is used in a filter.

### Does this PR introduce _any_ user-facing change?
No

### Is the change a dependency upgrade?
No

### How was this patch tested?
Unit tests and q test
 _mvn test -Dtest.output.overwrite -Dtest=TestMiniLlapLocalCliDriver -Dqfile=date_timestamp_partition_filter.q -pl itests/qtest -Pitests_

### CODE REVIEW

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreUtils.java [hunk 1]
```diff
@@ -74,25 +77,40 @@
...
+
+  public static String convertTimestampToString(Timestamp timestamp) {
+    return TIMESTAMP_FORMATTER.format(timestamp.toLocalDateTime());
+  }
+
+  public static Timestamp convertStringToTimestamp(String timestamp) {
+    LocalDateTime val = LocalDateTime.from(TIMESTAMP_FORMATTER.parse(timestamp));
+    return Timestamp.valueOf(val);
+  }
+
```
Comment (zabetak) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreUtils.java:113]: To avoid any kind of unpleasant surprises we should use the same methods and formatters in `PartFilterVisitor`.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/parser/ExpressionTree.java [hunk 2]
```diff
@@ -18,7 +18,7 @@
 package org.apache.hadoop.hive.metastore.parser;
 
 import java.sql.Timestamp;
-import java.util.Date;
```
Comment (zabetak) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/parser/ExpressionTree.java:21]: That's a sneaky change since `java.util.Date` is a super class of `java.sql.Date` so the behavior of `instanceof` and other operators may be affected. 

Do we have test coverage for this class? Are we creating anywhere else instances of `java.util.Date` that may end up here?

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+import java.sql.Date;
+import java.sql.Timestamp;
+import java.time.ZoneId;
+import java.util.TimeZone;
+
+import static org.junit.Assert.assertEquals;
+
+@Category(MetastoreUnitTest.class)
+public class TestMetaStoreUtils {
```
Comment (zabetak) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java:32]: Would some of these tests fail without the new changes introduced in this PR?

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+  @Test
+  public void testcConvertTimestampToString() {
+    TimeZone defaultTZ = TimeZone.getDefault();
+    try {
+      String date = MetaStoreUtils.convertTimestampToString(Timestamp.valueOf("2023-01-01 10:20:30"));
+      assertEquals("2023-01-01 10:20:30", date);
+    } finally {
+      TimeZone.setDefault(defaultTZ);
+    }
```
Comment (zabetak) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java:54]: Why do we need this? Neither the test nor the called method modify the default timezone.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java (hunk 3 above)
Comment (zabetak) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java:32]: Can we also add tests for:
* String -> Date -> String
* String -> Timestamp -> String

File: ql/src/test/queries/clientpositive/date_timestamp_partition_filter.q [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+INSERT into testpd(PartitionDate, col1, col2) VALUES('2023-01-02','Value21','Value22');
+explain extended select * from testpd where PartitionDate = '2023-01-01';
+select * from testpd where PartitionDate = '2023-01-01';
+
+
+CREATE EXTERNAL TABLE testpt(col1 string, col2 String) PARTITIONED BY(PartitionTimestamp TIMESTAMP) STORED AS ORC;
+INSERT into testpt(PartitionTimestamp, col1, col2) VALUES('2023-01-01 10:20:30','Value11','Value12');
+INSERT into testpt(PartitionTimestamp, col1, col2) VALUES('2023-01-02 20:30:40','Value21','Value22');
+explain extended select * from testpt where PartitionTimestamp = '2023-01-01 10:20:30';
+select * from testpt where PartitionTimestamp = '2023-01-01 10:20:30';
```
Comment (zabetak) [inline ql/src/test/queries/clientpositive/date_timestamp_partition_filter.q:14]: nit: Missing new line at the EOF.

File: ql/src/test/results/clientpositive/llap/date_timestamp_partition_filter.q.out [hunk 6]
```diff
@@ -0,0 +1,214 @@
+PREHOOK: query: CREATE EXTERNAL TABLE testpd(col1 string, col2 String) PARTITIONED BY(PartitionDate DATE) STORED AS ORC
```
Comment (zabetak) [inline ql/src/test/results/clientpositive/llap/date_timestamp_partition_filter.q.out:1]: Is there a part in this output that shows that partition pruning is working and there are no hidden exceptions that just prevent it from running?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/BaseSemanticAnalyzer.java [hunk 7]
```diff
@@ -1804,9 +1803,9 @@ private static String normalizeDateCol(Object colValue) throws SemanticException
...
     try {
-      return MetaStoreUtils.PARTITION_DATE_FORMAT.get().format(
-          MetaStoreUtils.PARTITION_DATE_FORMAT.get().parse(value.toString()));
-    } catch (ParseException e) {
+      return MetaStoreUtils.convertDateToString(
+          MetaStoreUtils.convertStringToDate(value.toString()));
+    } catch (Exception e) {
       throw new SemanticException(e);
     }
   }
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/BaseSemanticAnalyzer.java:1811]: Interestingly this seems to be dead code. We probably don't need to do something as part of this PR but I was wondering why it was left here. We can investigate later on after we are done with this.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreUtils.java (hunk 1 above)
Comment (mdayakar) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreUtils.java:113]: Modified `PartFilterVisitor` to use `MetaStoreUtils` APIs. Now both `PartFilterVisitor` and `MetaStoreDirectSql` are using same APIs.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/parser/ExpressionTree.java (hunk 2 above)
Comment (mdayakar) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/parser/ExpressionTree.java:21]: Actually in the flow `java.sql.Date `object only coming here, initially when added it was added wrongly as `java.util.Date`. As `java.util.Date` is super class for `java.sql.Date`, `value instanceof java.util.Date` is working fine till now.   `partition_date.q` test is covering it.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java (hunk 3 above)
Comment (mdayakar) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java:32]: Would some of these tests fail without the new changes introduced in this PR?
--> The issue was coming when SimpleDateFormat and java.sql.Date is used.

Can we also add tests for:

String -> Date -> String
String -> Timestamp -> String
--> Added the tests for the same.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java (hunk 4 above)
Comment (mdayakar) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java:55]: Actually here I missed out  `TimeZone.setDefault(TimeZone.getTimeZone(ZoneId.of("Asia/Hong_Kong")));` line in the try block. Thanks for pointing it out, now added in all the testcases.

File: ql/src/test/queries/clientpositive/date_timestamp_partition_filter.q (hunk 5 above)
Comment (mdayakar) [reply ql/src/test/queries/clientpositive/date_timestamp_partition_filter.q]: Added new line.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/BaseSemanticAnalyzer.java (hunk 7 above)
Comment (mdayakar) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/BaseSemanticAnalyzer.java:1811]: In case of `ParseException`, they were catching it and throwing `SemanticException`, now new code don't throw any exception in normal scenario so for the safer side catching `Exception` and throwing back `SemanticException`.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java (hunk 4 above)
Comment (zabetak) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java:55]: Since all tests are switching timezone it would be nice to do some refactoring to avoid the boilerplate. Using `@Before`, `@After` annotated methods and/or `@Parameterized` tests could aid in this.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/BaseSemanticAnalyzer.java (hunk 7 above)
Comment (mdayakar) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/BaseSemanticAnalyzer.java:1811]: Yeah, I could see only testcode is calling it. If required we can remove it(anyway as you mentioned its not related to the current PR).

File: ql/src/test/results/clientpositive/llap/date_timestamp_partition_filter.q.out (hunk 6 above)
Comment (mdayakar) [reply ql/src/test/results/clientpositive/llap/date_timestamp_partition_filter.q.out:1]: https://github.com/apache/hive/pull/4760/files#diff-a03353b02b81f6c2c36cc130d0d85b019fb725c7cb0d57d0baa9b7122177610eR41 and https://github.com/apache/hive/pull/4760/files#diff-a03353b02b81f6c2c36cc130d0d85b019fb725c7cb0d57d0baa9b7122177610eR90

These two lines shows that partition pruning is happening.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java (hunk 3 above)
Comment (mdayakar) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java:32]: > The issue was coming when SimpleDateFormat and java.sql.Date is used.
Let me rephrase my question cause it was not completely clear. Can someone see the problem after switching the formatter behind MetaStoreUtils.convertDateToString using the unit tests here?

Yes, the testcase fails when someone changes the formatter back to SimpleDateFormat. For example if I change like below
```java
public static String convertDateToString(Date date) {
    DateFormat val = new SimpleDateFormat("yyyy-MM-dd");
    val.setLenient(false); // Without this, 2020-20-20 becomes 2021-08-20.
    val.setTimeZone(TimeZone.getTimeZone("UTC"));
    return val.format(date);
    //return DATE_FORMATTER.format(date.toLocalDate());
  }
``` 
then the testcases are failing
```
[ERROR] Failures: 
[ERROR]   TestMetaStoreUtils.testConvertDateToString:46 expected:<202[3-01-0]1> but was:<202[2-12-3]1>
[ERROR]   TestMetaStoreUtils.testConvertStringToDateToString:72 expected:<202[3-01-0]1> but was:<202[2-12-3]1>
[INFO] 
[ERROR] Tests run: 6, Failures: 2, Errors: 0, Skipped: 0

```

Comment (mdayakar) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java:35]: > I am mostly interested for the string -> date/timestamp -> string case.

Added testcases for the same please check.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java (hunk 4 above)
Comment (mdayakar) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java]: Updated the testcases accordingly but I couldn't use `@Parameterized` due to some maven dependency issues.

File: ql/src/test/results/clientpositive/llap/date_timestamp_partition_filter.q.out (hunk 6 above)
Comment (mdayakar) [reply ql/src/test/results/clientpositive/llap/date_timestamp_partition_filter.q.out:1]: https://github.com/apache/hive/pull/4760/files#diff-a03353b02b81f6c2c36cc130d0d85b019fb725c7cb0d57d0baa9b7122177610eR41 and https://github.com/apache/hive/pull/4760/files#diff-a03353b02b81f6c2c36cc130d0d85b019fb725c7cb0d57d0baa9b7122177610eR90 shows partition pruning is happening but this is at storage level.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java (hunk 4 above)
Comment (zabetak) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java]: I pushed a commit (https://github.com/apache/hive/pull/4760/commits/4ff9ba968b31ea16611ba5d5bc05bbfb3b4fe184) based on what I had in mind. Have a look and let me know if I missed something.

With the new tests, I found a bug in `convertTimestampToString` using [zoneId=Europe/Paris, timestamp=9743-03-31T03:37:33] but I think this existed since day one so we can possibly defer it to another PR.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java (hunk 3 above)
Comment (mdayakar) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java:47]: Removed string -> date/timestamp -> string test cases as per comment in the commit (https://github.com/apache/hive/commit/4ff9ba968b31ea16611ba5d5bc05bbfb3b4fe184).

File: ql/src/test/results/clientpositive/llap/date_timestamp_partition_filter.q.out (hunk 6 above)
Comment (mdayakar) [reply ql/src/test/results/clientpositive/llap/date_timestamp_partition_filter.q.out:1]: Before fix the [output](https://github.com/apache/hive/pull/4760/files#diff-a03353b02b81f6c2c36cc130d0d85b019fb725c7cb0d57d0baa9b7122177610eR107) was empty, after fix it is returning the corresponding partition matching rows.

Comment (zabetak) [reply ql/src/test/results/clientpositive/llap/date_timestamp_partition_filter.q.out:1]: The query result is definitely useful, I am asking for the EXPLAIN output.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java (hunk 4 above)
Comment (zabetak) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java]: Yeap I agree its not a problem introduced by the PR but it is a bug that is probably due to DST shift. We have many of those in Hive :) Please log a new JIRA ticket and we can follow-up there.

File: ql/src/test/results/clientpositive/llap/date_timestamp_partition_filter.q.out (hunk 6 above)
Comment (mdayakar) [reply ql/src/test/results/clientpositive/llap/date_timestamp_partition_filter.q.out:1]: There is no change in the EXPLAIN plan output, I just added to make sure partition filter pushdown is happening properly.

**Review by zabetak — APPROVED** [review]

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java (hunk 4 above)
Comment (mdayakar) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java]: Issue is observed for the timestamps from `9743-03-31 03:00:00` to `9743-03-31 03:59:59`.

Comment (zabetak) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java]: https://issues.apache.org/jira/browse/HIVE-27775 is the ticket for the problem reported here.
